addons/web/controllers/main.py

```diff
@@ -924,6 +924,21 @@ def web_login(self, redirect=None, **kw):
         response.headers['X-Frame-Options'] = 'DENY'
         return response
 
+    # sid changes once connected, therefore csrf changes too.
+    @http.route('/web/accept_scope', type='http', auth="user", methods=['GET'], website=True)
+    def accept_scope(self, **values):
+        return request.render('web.app_auth', values)
+
+    @http.route('/web/oauth', type='http', auth="user", methods=['POST'])
+    def web_oauth(self, redirect=None, scope=None, do="refuse", **kw):
+        if do == 'accept':
+            token = request.env['res.users.apikeys']._generate(scope, scope) #name of the token, and the actual scope
+            redirect = "%s?success=1&token=%s" % (redirect, token)
+        else:
+            redirect = redirect + "?success=0"
```

Comment:
Should use werkzeug.urls to parse and augment the URL: these will break if `redirect` already has query parameters.